tasks.py

```diff
@@ -1,84 +1,105 @@
-    """
-
-    with open("tasks.csv", "a") as tasks_file:
-        writer = csv.writer(tasks_file)
-        writer.writerow([name, False])
-
-
-def complete():
-    """
-    Mark an existing task as completed.
-    """
-
-    with (
-        open("tasks.csv") as tasks_file,
-        NamedTemporaryFile("w", delete=False) as new,
-    ):
-        reader = csv.reader(tasks_file)
-        print("Current tasks:")
-        for id, (name, completed) in enumerate(reader):
-            print(id, name, completed)
-
-        to_complete = int(input("task ID?> "))
-        writer = csv.writer(new)
-        tasks_file.seek(0)
-        for id, (name, completed) in enumerate(reader):
-            if id == to_complete:
-                writer.writerow([name, True])
-            else:
-                writer.writerow([name, completed])
-
-    os.rename(new.name, "tasks.csv")
-
-
-operations = dict(
-    create=create,
-    complete=complete,
-    list=list,
-)
-
-
-def main():
-    print("Enter a command [create, list, complete].")
-    while True:
-        line = input("-> ").strip()
-        if not line:
-            return
-
-        operation, *args = line.split()
-        fn = operations.get(operation)
-        fn(*args)
-
-
-if __name__ == "__main__":
-    main()
+#!/usr/bin/env python3
+"""
+A simple task manager.
+For the purposes of our exercise we store tasks in a file which we'll
+call tasks.csv (a CSV file). We'll talk more later about other ways to
+store state which are more robust.
+"""
+from tempfile import NamedTemporaryFile
+import datetime
+import csv
+import os
+import sys
+import pandas as pd
+
+import collections
+
+try:
+    collectionsAbc = collections.abc
+except AttributeError:
+    collectionsAbc = collections
+
+DONE = " ✅"
+
+
+def list(stdout=sys.stdout, filename="tasks.csv"):
+    """
+    List the current known tasks.
+    """
+
+    with open(filename) as tasks_file:
+        reader = csv.reader(tasks_file)
+        for name,format_date,completed in reader:
+            stdout.write(f"{name,format_date}{DONE if eval(completed) else ''}\n")
+
+
+def create(name, filename="tasks.csv"):
+    """
+    Create a new task.
+    """
+
+    with open(filename, "a", newline="") as tasks_file:
+        writer = csv.writer(tasks_file)
+        print('Task successfully created. Please enter a date in the MM-DD-YY format:')
+        strdate = input()
+        format_date = datetime.datetime.strptime(strdate,'%m-%d-%Y')
+        writer.writerow([name, format_date, False])
+
+
+def complete(stdout=sys.stdout,filename="tasks.csv"):
+    """
+    Mark an existing task as completed.
+    """
+
+    with (
+        open("tasks.csv") as tasks_file,
+        NamedTemporaryFile("w", delete=False) as new,
+    ):
+        reader = csv.reader(tasks_file)
+        print("Current tasks:")
+        for id, (name, format_date, completed) in enumerate(reader):
+            print(id+1, name, format_date, DONE if eval(completed) else '')

```

Comment:
Better to generally avoid `eval` -- here you could use `if completed == "True"`, or otherwise change what's put in the file, say to be `"x"` when completed, and empty string (`""`) for not completed perhaps.

But eval can be a minor security risk -- it allows someone who can provide untrusted input to potentially run arbitrary code -- I'll elaborate more on what that means in class in a follow-up lecture, but here's an article you may find interesting: 

https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html